dll/ntdll/rtl/hashtable.c

```diff
@@ -0,0 +1,1264 @@
+    {
+        LIST_ENTRY *EntryToBeMoved = RemoveHeadList(ChainToBeMoved);
+        RTL_DYNAMIC_HASH_TABLE_ENTRY *HashEntryToBeMoved = RtlFlinkToHashEntry(&EntryToBeMoved->Flink);
+
+        while (CurEntry->Flink != CombinedChain)
+        {
+            LIST_ENTRY *NextEntry = CurEntry->Flink;
+            RTL_DYNAMIC_HASH_TABLE_ENTRY *NextHashEntry = RtlFlinkToHashEntry(&NextEntry->Flink);
+
+            if (NextHashEntry->Signature >= HashEntryToBeMoved->Signature)
+            {
+                break;
+            }
+
+            CurEntry = NextEntry;
+        }
+
+        InitializeListHead(CurEntry, &(HashEntryToBeMoved->Linkage));
+    }
+
+    // Finally free any extra memory if possible.
+    uint32_t FirstLevelIndex = 0, SecondLevelIndex;
+    RtlComputeDirIndices(HashTable->TableSize, &FirstLevelIndex, &SecondLevelIndex);
+
+    if (SecondLevelIndex == 0)
+    {
+        LIST_ENTRY **FirstLevelDir = HashTable->FirstLevelDir;
+        LIST_ENTRY *SecondLevelDir = FirstLevelDir[FirstLevelIndex];
+
+        ExFreePoolWithTag(SecondLevelDir, TAG_HASHTABLE_MEMBER);
+
+        FirstLevelDir[FirstLevelIndex] = NULL;
+
+        // Switch to a single-dir mode if fits within a single second-level.
+        if (HT_SECOND_LEVEL_DIR_MIN_SIZE == HashTable->TableSize)
+        {
+            HashTable->SecondLevelDir = FirstLevelDir[0];
+            ExFreePoolWithTag(FirstLevelDir, TAG_HASHTABLE_MEMBER);
+        }
+    }
+
+    return TRUE;
+}
+
+#pragma endregion
+
+_Must_inspect_result_
+NTSYSAPI
+BOOLEAN
+NTAPI
+RtlCreateHashTable(
+    _Inout_ _When_(NULL == *HashTable, __drv_allocatesMem(Mem)) RTL_DYNAMIC_HASH_TABLE **HashTable,
+    _In_ ULONG Shift,
+    _In_ _Reserved_ ULONG Flags)
+{
+    DPRINT1("RtlCreateHashTable() called\n");
+
+    uint16_t InitialSize = RTL_DYNAMIC_HASH_TABLE_MIN_SIZE;
+
+    // TODO: If it is 128 all the time, why there is a check?
+    // There must be another function, probably allowing InitialSize as a parameter, and for
+    // undocumented internal use. By default, it is 128 here. When the other data structure is
+    // discovered, we can create the new internal function. Then we can push all this code into
+    // internal function, with RTL_DYNAMIC_HASH_TABLE_MIN_SIZE as parameter.
+
+    // Initial size must be a power of two and within the allowed range.
+    // if (!IS_POWER_OF_TWO(InitialSize) ||
+    //     (InitialSize > MAX_HASH_TABLE_SIZE) ||
+    //     (InitialSize < BASE_HASH_TABLE_SIZE)) {
+    //     return FALSE;
+    //     }
+
+    // First allocate the hash table header.
+    uint32_t LocalFlags = 0;
+    RTL_DYNAMIC_HASH_TABLE **Table;
+    if (*HashTable == NULL)
+    {
+        Table = ExAllocatePoolWithTag(NonPagedPoolNx, sizeof(RTL_DYNAMIC_HASH_TABLE), TAG_HASHTABLE);
+        if (Table == NULL)
+        {
+            DPRINT1(
+                "Allocation of '%s' failed. (%llu bytes)", "RTL_DYNAMIC_HASH_TABLE", sizeof(RTL_DYNAMIC_HASH_TABLE));
```

Comment:
```suggestion
            DPRINT1("Alloc '%s' failed. %llu bytes", "RTL_DYNAMIC_HASH_TABLE", sizeof(RTL_DYNAMIC_HASH_TABLE));
```

Reply:
That thing is due to VS Code auto-format on save. I fixed many issues but not all. I updated it and soon will commit & squash & force-push here with other changes.

Reply:
Then I guess, disable the auto-format feature, or use another editor that doesn't have that thing enabled by default

Reply:
Hi. I disabled it after your comment yesterday and formatted it manually yet it's relatively long wall of text and things like these are easy to miss.